pkg/gatewayserver/grpc_nsgs.go

```diff
@@ -82,7 +82,8 @@ func (gs *GatewayServer) ScheduleDownlink(ctx context.Context, down *ttnpb.Downl
 		}
 		ctx = events.ContextWithCorrelationID(ctx, events.CorrelationIDsFromContext(conn.Context())...)
 		down.CorrelationIDs = append(down.CorrelationIDs, events.CorrelationIDsFromContext(ctx)...)
-		registerSendDownlink(ctx, conn.Gateway(), down, conn.Frontend().Protocol())
+		msg := *down
+		registerSendDownlink(ctx, conn.Gateway(), &msg, conn.Frontend().Protocol())
```

Comment:
What this does is only copy `ttnpb.DownlinkMessage`, including pointers to the same memory space, so this doesn't resolve the issue.

---

10 lines above it says that `down` will be owned by the connection. I think that the actual issue is that `down` isn't actually owned by the connection: it is being used after `ScheduleDown()`, like here.

So I'd suggest renaming `down` on L75 to `connDown` and use that on L76-L77, and don't change anything else.